Restore coverage for Marketplace tests on FxOS

bobsilverberg opened the pull request:
This PR includes a number of changes which should fix most of the tests, and xfails a couple of others.

This addresses issue #114 and includes the xfail mentioned in issue #115, it also includes some refactoring as mentioned in issue #96 

Among the changes:
Use first paid app rather than a specific app
Fix back button locator

bobsilverberg commented:
@davehunt r? please and thank you

davehunt commented on marketplacetests/marketplace/pages/base.py:
```diff
@@ -36,11 +35,13 @@ def login(self, username, password):
+        name = app.name
+        author = app.author
+        self.tap_home()
+        return {'name': name, 'author': author}
```
I'm guessing we're not returning the app object because the properties will be accessed when it's no longer available? I don't think we need to have `name` and `author` variables, we could just: `return {'name': app.name, 'author': app.author}`. Or, if this is all we're returning we could return multiple values: `return app.name, app.author`. The former has better structure and readability though. \</rambling>

bobsilverberg replied:
The problem is that the call to `self.tap_home()` causes the variable `app` to lose its context, so we cannot call `app.name` after that, which is why I stored the values in a variable. I guess I could just create the dictionary on line 41 and then return that on line 44, which would save a line of code. I'll do that.

bobsilverberg replied:
I also considered returning multiple values, but most methods that call this only care about the name, and not the author, so the dictionary seemed better.

davehunt commented on marketplacetests/tests/test_marketplace_create_confirm_pin.py:
```diff
@@ -13,22 +13,18 @@ class TestMarketplaceCreateConfirmPin(MarketplaceGaiaTestCase):
         home_page = marketplace.launch()
 
-        home_page.login(acct.email, acct.password)
+        app_name = home_page.find_first_paid_app()
```
Calling this causing us to perform the search twice, the second search is by app name (are we certain this will be unique?). Could we just perform the `:paid` search within the test?

bobsilverberg replied:
I was not really happy with this either, but it seemed necessary based on everything else that was happening behinf the scenes (e.g., checking for the app to be installed and logging in). Your comment made me realize that I missed something, which is that just because marionette loses its context doesn't mean we have to reload the page - we just have to find the element again!  I am now working on a better version of this PR which I will submit for another review shortly.

davehunt commented on marketplacetests/tests/test_marketplace_forgot_pin.py:
```diff
@@ -14,23 +14,19 @@ class TestMarketplaceForgotPin(MarketplaceGaiaTestCase):
         home_page = marketplace.launch()
 
-        home_page.login(acct.email, acct.password)
+        app_name = home_page.find_first_paid_app()
```
As above, searching just once would be much more efficient, unless there's a reason for multiple searches? If there is, perhaps the first could be via the API or headless request? (I'm fine with this being addressed separately).

davehunt commented:
Looks good, however I'd like to see if we can avoid searching twice as I suspect this is rather costly. Also, I haven't run this patch locally, but I will once my comments are addressed.

bobsilverberg commented:
@davehunt Thanks for the review. I added a new commit that I think is a huge improvement over the initial one. Could you find it in the goodness of your heart to grant me another review?

bobsilverberg commented on marketplacetests/tests/test_marketplace_search_for_paid_app.py:
```diff
@@ -10,25 +10,13 @@ class TestSearchMarketplacePaidApp(MarketplaceGaiaTestCase):
 
     def test_search_paid_app(self):
```
Note that I confirmed the change to this test with @krupa as it is not functionally equivalent to its predecessor. She said the new version of the test is fine.

davehunt commented on marketplacetests/tests/test_marketplace_search_for_paid_app.py:
```diff
@@ -10,25 +10,13 @@ class TestSearchMarketplacePaidApp(MarketplaceGaiaTestCase):
-        self.fail('The app: %s was not found.' % app_name)
+        saved_price = app.install_button_text
+        details_page = app.tap_app()
+        self.assertEqual(saved_price, details_page.install_button_text)
```
This doesn't check if the app is actually a paid app, does it? If the button text on both results and details pages was 'Free' then the test would still pass. This isn't related to your change, but possibly we can enhance this test in a separate patch?

bobsilverberg replied:
Good idea. Do you want to open an issue for that? ;)

davehunt replied:
Perhaps @krupa can comment and open an issue if it's needed?

davehunt commented on marketplacetests/marketplace/pages/base.py:
```diff
@@ -137,6 +129,23 @@ def show_menu(self):
             lambda m: 'mkt-nav--visible' in body.get_attribute('class'))
         return NavMenu(self.marionette)
 
+    def tap_install_button_of_first_paid_app(self):
```
I don't find it particularly intuitive that this method results a page and an app name (we would at least need a docstring). I think I would prefer something like:

```python
def search_for_paid_apps(self):
    self.set_region('us')
    results_page = self.search(':paid')
    return results_page

def is_app_installed(self, name):
    installed = self.apps.is_app_installed(name)
    self.switch_to_marketplace_frame()
    return installed
```

```python
def test_method(self):
    acct = FxATestAccount(base_url=self.base_url).create_account()
    marketplace = Marketplace(self.marionette, self.MARKETPLACE_DEV_NAME)
    home_page = marketplace.launch()
    home_page.login(acct.email, acct.password)
    app = home_page.search_for_paid_apps().search_results[0]
    self.assertFalse(page.is_app_installed(app.name))
    app.tap_install_button()
    ...
```

bobsilverberg replied:
I agree that it is less than intuitive, but going with your suggestion re-adds much of the duplication I was trying to remove. I don't like the fact that each test needs to:
- locate a paid app
- verify that the app isn't installed
- tap on the install button for the app

and that we need to include each of those in the test. I was trying to encapsulate all of that into a single helper method, but the problem is that we need the app name in the test afterwards, and we also end up at the SearchResults page, so I felt it made sense that our helper return that.

I will try a different version where the helper is a test method, rather than a page object method, which might be a bit better. If that is still not good then we can go with what you suggest above.


Seems

bobsilverberg commented:
@davehunt I added a new commit which uses a base test method instead of a page object. Still not ideal, but you'll see in there that we need even more code than in your example for `test_method` because of losing context, so I'd really like to find a way to encapsulate all of these steps into a method instead of having a ton of duplicate code in each test.

I await your next round of comments. :)

davehunt commented on marketplacetests/marketplace/pages/base.py:
```diff
@@ -137,6 +129,18 @@ def show_menu(self):
+        self.set_region('us')
+        return self.search(':paid')
+
+    def is_app_installed(self, name):
```
Thinking about it, this probably also belongs to `MarketplaceGaiaTestCase` as it doesn't use the page/app object.

bobsilverberg replied:
It uses `switch_to_marketplace_frame` which exists in the app object, so by default it will not have access to that if we put it into `MarketplaceGaiaTestCase`. For that reason I would prefer to leave it as is.

davehunt replied:
Ah right, I missed that. Doesn't seem ideal but I can live with it. :)

davehunt commented on marketplacetests/marketplace_gaia_test.py:
```diff
@@ -76,3 +76,18 @@ def install_certs(self):
         self.wait_for_element_not_displayed(*self._statusbar_downloads_icon_locator)
+
+    def tap_install_button_of_first_paid_app(self, page):
+        """This method taps on the install button of the first paid app
```
The summary should be a single line. See http://legacy.python.org/dev/peps/pep-0257/#multi-line-docstrings

davehunt commented on marketplacetests/marketplace_gaia_test.py:
```diff
@@ -76,3 +76,18 @@ def install_certs(self):
         self.marionette.switch_to_frame()
         self.wait_for_element_not_displayed(*self._statusbar_downloads_icon_locator)
+
+    def tap_install_button_of_first_paid_app(self, page):
```
I wonder if we need the page or if we can just instantiate a base page from within this method?

bobsilverberg replied:
Not requiring the caller to pass in a page makes sense. I'll need to make a change to the `BasePage` object in order to make it work, but that seems fine. I'll give it a try.

davehunt commented on marketplacetests/marketplace_gaia_test.py:
```diff
@@ -76,3 +76,18 @@ def install_certs(self):
+        self.app_name = search_results_page.search_results[0].name
+        self.assertFalse(page.is_app_installed(self.app_name))
+        # We need to re-find the app object as the call to is_app_installed loses context
+        search_results_page.search_results[0].tap_install_button()
```
I see, that makes sense.

davehunt commented:
This looks good, just a few minor comments but r+ with those addressed. Note that I haven't run the tests locally. If you'd like me to please let me know.

bobsilverberg commented:
@davehunt I addressed your final comments and squashed the commits. I will wait for Travis to say it's okay and then I'll merge this to restore coverage. Thanks for the reviews.

bobsilverberg commented:
No word from Travis in over 2 hours. That's weird. I'm going to just go ahead and merge this anyway.

davehunt commented:
Looks like this introduced a flake8 issue:

> ./marketplacetests/tests/test_marketplace_login_during_purchase.py:21:9: F841 local variable 'home_page' is assigned to but never used

davehunt commented:
> No word from Travis in over 2 hours. That's weird. I'm going to just go ahead and merge this anyway.

We should be able to retrigger Travis jobs if they're stalled. Alternatively, while Travis is just running flake8, we can do this locally (it's good to set up a pre-commit hook).
